scRNA-seq/06-overrepresentation_analysis.Rmd

```diff
@@ -0,0 +1,414 @@
+
+```{r filter_padj, live = TRUE}
+go_result_df %>%
+  dplyr::filter(p.adjust < 0.00001)
+```
+
+Some of these columns have names that are somewhat opaque and are used for visualization, so let's briefly talk about them here!
+
+| Column name | Numerator | Denominator |
+|-------------|-----------|-------------|
+| `GeneRatio` | Number of genes in the top 100 marker genes and the specific gene set (also in `Count` column) | Number of genes in the top 100 marker genes and _any_ gene set |
+| `BgRatio`   | Number of genes in the background set and the specific gene set | Number of genes in the background set and _any_ gene set |
+
+#### Visualizing results
+
+We can use a dot plot to visualize our significant enrichment results.
+
+```{r dotplot, live = TRUE}
+enrichplot::dotplot(go_ora_results)
+```
+
+**What cell type do you think the cells from cluster 1 are?**
+
+We can use an [UpSet plot](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4720993/) to visualize the **overlap** between the gene sets that were returned as significant.
+
+```{r upsetplot, live = TRUE}
+enrichplot::upsetplot(go_ora_results)
+```
+
+It's important to keep in mind that gene sets or pathways aren't independent, either!
+This is particularly true of GO terms, which are hierarchical in nature.
+
+### CellMarker ORA
+
+Next, we'll use another resource, CellMarker, as additional support for our conclusions based on GO.
+
+#### About CellMarker
+
+[CellMarker](http://bio-bigdata.hrbmu.edu.cn/CellMarker/) is a manually curated database of cell marker data for human and mouse ([Zhang *et al.* 2018](https://doi.org/10.1093/nar/gky900)).
+Studies/evidence of the following nature were included in the manual curation process ([Zhang *et al.* 2018](https://doi.org/10.1093/nar/gky900)):
+
+* scRNA-seq
+* Flow cytometry
+* Immunostaining
+
+And there are both gene and protein markers.
+
+Let's read in the CellMarker data we've already cleaned:
+
+```{r read_in_cm, live = TRUE}
+cellmarker_df <- readr::read_tsv(cellmarker_file)
+```
+#### Run `enricher()`
+
+`clusterProfiler::enricher()` is a more general way to perform ORA -- we specify the pathways or gene sets ourselves.
+We can use the it to run ORA with the CellMarker data, which contains gene symbols.
+That's why we needed to perform gene identifier conversion earlier!
+
+```{r cm_ora, live = TRUE}
+cellmarker_ora_results <- enricher(
+  gene = cluster01_markers,  # Genes of interest
+  pvalueCutoff = 0.05,  # More permissive cut off
+  pAdjustMethod = "BH",  # FDR
+  universe = background_set,  # Background set
+  # The pathway information should be a data frame with a term name or 
+  # identifier and the gene identifiers
+  TERM2GENE = cellmarker_df
+)
+```
+
+#### Visualizing results
+
+Let's look at a bar plot of the significant results.
+
+```{r cm_barplot, live = TRUE}
+# You can use base graphics barplot() with the ORA results object
+barplot(cellmarker_ora_results)
```

Comment:
So it turns out this is overloading barplot, but using ggplot! 🙄 I discovered this because I was annoyed that there was no label for the x axis, and adding `xlab` (or `ylab`) to the function call didn't work. But this does:
```
barplot(cellmarker_ora_results) +
  ggplot2::xlab("count")
```

I don't know what to do with this behavior, because it is bad! Don't use the base `barplot` function if you are not using base graphics! But I also hate that there is no label on the axis.

Reply:
I think if I don't make this live, I can not use the helper comment that might be misleading. I was using `enrichplot::barplot()` (not a thing!) before and I think the temptation to do that here will persist without saying something about "base."

Reply:
Updated in https://github.com/AlexsLemonade/training-modules/pull/483/commits/fa652aa78c74698b7b9bf3b0c0bf5846d1b849ec

Reply:
Looks good enough to me. 

But I still don't approve of the behavior... also that `enrichplot::barplot` is not a thing, nor is `enrichplot:::barplot()`, which makes me realize that I know less than I thought about how S3 objects and overloading works.